outlines/types/__init__.py

```diff
@@ -1,4 +1,31 @@
-from . import airports, countries
-from .email import Email
-from .isbn import ISBN
-from .locales import locale
+from enum import Enum
+
+from . import airports, countries, locale
+from .dsl import Regex, json_schema, one_or_more, optional, regex, repeat, zero_or_more
+
+# Python types
+integer = Regex(r"[+-]?(0|[1-9][0-9]*)")
+boolean = Regex("(True|False)")
+number = Regex(rf"{integer.pattern}(\.[0-9]+)?([eE][+-][0-9]+)?")
+date = Regex(r"(\d{4})-(0[1-9]|1[0-2])-([0-2][0-9]|3[0-1])")
+time = Regex(r"([0-1][0-9]|2[0-3]):([0-5][0-9]):([0-5][0-9])")
+datetime = Regex(rf"({date.pattern})(\s)({time.pattern})")
+
+# Basic regex types
+digit = Regex(r"\d")
+char = Regex(r"\w")
+newline = Regex(r"(\r\n|\r|\n)")  # Matched new lines on Linux, Windows & MacOS
+whitespace = Regex(r"\s")
+
+# Document-specific types
+sentence = Regex(r"[A-Z].*\s*[.!?]")
+paragraph = Regex(rf"{sentence.pattern}(?:\s+{sentence.pattern})*\n+")
+
+
+# Custom types
```

Comment:
Would be nice to have specification, references or sources that tell where does complex Regexes come from!

Reply:
I infer that `email` have been taken from Stack Overflow https://stackoverflow.com/a/201378/14773537

Reply:
IDEM, that would be nice to not lose that comment about ISBN:

>  Matches any ISBN number.  Note that this is not completely correct as not all
> 10 or 13 digits numbers are valid ISBNs. See https://en.wikipedia.org/wiki/ISBN
> Taken from O'Reilly's Regular Expression Cookbook:
> https://www.oreilly.com/library/view/regular-expressions-cookbook/9781449327453/ch04s13.html
> TODO: Can this be represented by a grammar or do we need semantic checks?